### PR TITLE
eip7732: add process execution payload header tests

### PR DESCRIPTION
**Pull request by terencechain**

### CODE REVIEW

**Comment by terencechain** [comment]
thanks a lot! ill remember to run make lint next time

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+
+def make_validator_builder(spec, state, validator_index):
+    """
+    Helper to make a validator a builder by setting builder withdrawal credentials
+    """
+    # Set builder withdrawal credentials (0x03 prefix)
+    builder_credentials = spec.BLS_WITHDRAWAL_PREFIX + b"\x00" * 31
+    builder_credentials = spec.BUILDER_WITHDRAWAL_PREFIX + builder_credentials[1:]
+    state.validators[validator_index].withdrawal_credentials = builder_credentials
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: For consistency, we should probably mirror these functions:

https://github.com/ethereum/consensus-specs/blob/56899c5dbb8511c4fe10ec05d0b2afd1159882fb/tests/core/pyspec/eth2spec/test/helpers/withdrawals.py#L127-L151

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    # Non-builder but not the proposer
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec, state, builder_index=other_index, value=spec.Gwei(0)
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
+
+
+#
+# Balance validation tests
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:434]: Another balance test I'd like to see:

* There is a builder with an excess balance, eg 2048.25 ETH.
* It is able to submit a bid for 2016.25 ETH.

Sort of an edge case because the bid limit depends on the builder's balance, not its effective balance.

Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:434]: added in #[4f2db69](https://github.com/ethereum/consensus-specs/pull/4457/commits/4f2db6950780bf31f4b53c922d0c7f8c6ce9053c)

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py (hunk 1 above)
Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: modified in [4f2db69](https://github.com/ethereum/consensus-specs/pull/4457/commits/4f2db6950780bf31f4b53c922d0c7f8c6ce9053c)

**Review by jtraglia — APPROVED** [review]

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+    block_hash=None,
+    blob_kzg_commitments_root=None,
+    valid_signature=True,
+):
+    """
+    Helper to create a signed execution payload header with customizable parameters.
+    Note: Always call build_empty_block_for_next_slot BEFORE this function if using it,
+    since that function advances the state slot.
+    """
+    if builder_index is None:
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: If this function takes a slot as parameter, don't we need to advance the state?. Moreover if we do this, then it's not necessary to have the comment of `build_empty_block_for_next_slot` but rather just specify that if `slot==None` then the current state slot will be used.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+        builder_index = spec.get_beacon_proposer_index(state)
+
+    if slot is None:
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:56]: I think I would use the logic:

```
    if slot is None:
        slot = state.slot
    assert slot >= state.slot
    process_slots(state, slot)
    ...

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+        fee_recipient = spec.ExecutionAddress()
+
+    if gas_limit is None:
+        gas_limit = spec.uint64(30000000)
+
+    if block_hash is None:
+        block_hash = spec.Hash32()
+
+    if value is None:
+        value = spec.Gwei(0)
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:80]: We need to also check that if value is not 0 then builder index equals proposer index

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+
+    return spec.SignedExecutionPayloadHeader(
+        message=header,
+        signature=signature,
+    )
+
+
+def make_validator_builder(spec, state, validator_index):
+    """
+    Helper to make a validator a builder by setting builder withdrawal credentials
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]:
```suggestion
    Helper to make a validator a builder by setting builder withdrawal credentials.
```

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+    """
+    Test valid builder scenario with registered builder and non-zero value
+    """
+    builder_index = spec.get_beacon_proposer_index(state)
+
+    # Make the validator a registered builder
+    make_validator_builder(spec, state, builder_index)
+
+    # Ensure builder has sufficient balance
+    value = spec.Gwei(1000000)  # 1 ETH
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: Two comments:
1000000 = 0.001 ETH != 1 ETH

Also I don't understand how this test can pass: it's a self builder with a non-zero value.

Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: In this case, the proposer is also a builder. It has 0x03

```
builder_index = spec.get_beacon_proposer_index(state)
make_validator_builder(spec, state, builder_index)
```

Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: I made sure to have all 4 cases now:

1. (test_process_execution_payload_header_non_builder_non_zero_value): Self builder (no 0x03) betting non-zero should fail
2. (test_process_execution_payload_header_valid_self_build): Self builder (no 0x03) bidding zero should pass
3. (test_process_execution_payload_header_valid_builder_zero_value): Builder (0x03) bidding zero should pass (newly added)
4. (test_process_execution_payload_header_valid_builder): Builder (0x03) bidding non-zero should pass

**Comment by terencechain** [comment]
Please merge https://github.com/ethereum/consensus-specs/pull/4464 before this one, I will have to update the tests

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py (hunk 5 above)
Comment (jtraglia) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:80]: Yeah we're still missing this check @terencechain.

Needs something like:

```python
if value == 0 && builder_index != proposer_index:
   raise ...
```

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+    builder_index = spec.get_beacon_proposer_index(state)
+    state.validators[builder_index].exit_epoch = spec.get_current_epoch(state)
+
+    make_validator_builder(spec, state, builder_index)
+
+    block, signed_header = prepare_block_with_execution_payload_header(
+        spec, state, builder_index=builder_index, value=spec.Gwei(1000000)
+    )
+
+    yield from run_execution_payload_header_processing(spec, state, block, valid=False)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: I don't believe this would work even if the builder were active. It doesn't have an excess balance, right? So it couldn't make a bid.

Comment (jtraglia) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: Same thing for other tests below.

* `test_process_execution_payload_header_slashed_builder`
* `test_process_execution_payload_header_non_builder_non_zero_value`

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py (hunk 5 above)
Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:80]: it's right below line 89-90 right?

https://github.com/ethereum/consensus-specs/pull/4457/files#diff-076f15fc245573c84a9558019cca1ec66ce3f9538edff84d5cbe3c7b5fefca4cR89-R90

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py (hunk 8 above)
Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: ah, yes, but active and slashed checks happen before balance check
but you are right I should have good balance regardless

Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: added in [fc31e05](https://github.com/ethereum/consensus-specs/pull/4457/commits/fc31e05e082d53936991c5ec0df1d8f617e2257d)

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py (hunk 5 above)
Comment (jtraglia) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:80]: Hmm I suppose it does 👍

Comment (potuz) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:80]: I think we need to make this an if and only if, see my below comment in ` test_process_execution_payload_header_valid_builder`

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+
+
+@with_eip7732_and_later
+@spec_state_test
+@always_bls
+def test_process_execution_payload_header_valid_builder(spec, state):
+    """
+    Test valid builder scenario with registered builder and non-zero value
+    """
+    builder_index = spec.get_beacon_proposer_index(state)
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: This makes a self-proposing builder. I think we want to test with builders that aren't self-proposing.

The following assertion is in `process_execution_payload_header` 
```    
    if builder_index == block.proposer_index:
        assert amount == 0
```
How can this test pass?

Also a separate discussion that we may need to have is about the value for non-self-builds being forced to be non-zero or avoid adding a withdrawal for a zero value.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+    """
+    yield "pre", state
+    yield "block", block
+
+    if not valid:
+        expect_assertion_error = True
+    else:
+        expect_assertion_error = False
+
+    if expect_assertion_error:
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: Why the extra variable `expect_assertion_error`  when it's always `!valid`? Mr Claude is playing logic tricks.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+
+    # Make the proposer a registered builder
+    make_validator_builder(spec, state, proposer_index)
+
+    # Ensure builder has sufficient balance
+    value = spec.Gwei(1000000)  # 0.001 ETH
+    required_balance = value + spec.MIN_ACTIVATION_BALANCE
+    state.balances[proposer_index] = required_balance
+
+    block = build_empty_block_for_next_slot(spec, state)
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:362]: Why this test prepares explicitly the header and block instead of using the helper? notice that this test is exactly the same scenario of the one above that I remarked that it shouldn't pass, and indeed in this one we assert that processing fails, but in the other we assert that it succeeds.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py (hunk 9 above)
Comment (terencechain) [reply tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: it's actually different
in a given block, builder index is proposer at slot N, and proposer index is slot N+1
```
builder_index = spec.get_beacon_proposer_index(state) # this is builder index is proposer at slot N
block, signed_header = prepare_block_with_execution_payload_header(
        spec, state, builder_index=builder_index, value=value
    ) # this builds block for slot N+1
```
but I understand this is confusing to read, so its better to change the code to not use `builder_index = spec.get_beacon_proposer_index(state)`

**Comment by terencechain** [comment]
> Tests are always for self-building, when the builder is actually a 0x03 I would value more if they weren't self building.

it's not always self building, `builder_index` is the parent slot proposer index instead of intended slot, but I will fix them to read better

```python
   block, signed_header = prepare_block_with_execution_payload_header(
        spec, state, builder_index=builder_index, value=value
    )
```
 
> Tests that missing (besides all that have different builder than proposer)

I will add them now

**Comment by terencechain** [comment]
@potuz all of your comments from today is addressed in addb6b1981bc96195915c04c454846944396d7cc

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 12]
```diff
@@ -472,6 +524,229 @@ def test_process_execution_payload_header_excess_balance(spec, state):
...
+@with_eip7732_and_later
+@spec_state_test
+def test_process_execution_payload_header_sufficient_balance_with_pending_payments(spec, state):
+    """
+    Test builder with sufficient balance for both bid and existing pending payments
+    """
+    block, builder_index = prepare_block_with_non_proposer_builder(spec, state)
+
+    # Set up scenario: balance=2000, bid=600, existing_pending=500, min_activation=32ETH
+    # Total needed: 600 + 500 + 32000000000 = ~32.0011 ETH < 2000 ETH (should pass)
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py:579]: Why does the previous test not account for the 32ETH minimum balance?

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py [hunk 13]
```diff
@@ -472,6 +524,229 @@ def test_process_execution_payload_header_excess_balance(spec, state):
...
+    spec, state
+):
+    """
+    Test builder with sufficient balance for bid alone but insufficient when considering pending withdrawals
+    """
+    block, builder_index = prepare_block_with_non_proposer_builder(spec, state)
+
+    # Set up scenario: balance=1000, bid=600, existing_withdrawal=500
+    # Total needed: 600 + 500 = 1100 > 1000 (should fail)
+    balance = spec.Gwei(1000)
```
Comment (potuz) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_execution_payload_header.py]: Same here, why aren't we accounting for the 32 ETH in this check?
```
    assert (
        amount == 0
        or state.balances[builder_index]
        >= amount + pending_payments + pending_withdrawals + MIN_ACTIVATION_BALANCE
    )
```

**Review by potuz — CHANGES_REQUESTED** [review]
The failing tests fail because of the MIN_ACTIVATION_BALANCE rather than the pending payments. 

TODO: We should revisit these tests if this condition is changed in the spec.
